### PR TITLE
Add foreign function interface for PARSEC [MAID-2942, MAID-2955]

### PR DESCRIPTION
**Pull request by nbaksalyar**
* Add FFI functions to be used from languages other than Rust.
* Make integration tests more generic and implementation-independent (for testing FFI).

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -46,10 +54,64 @@ quick_error! {
...
+            Error::MismatchedPayload => ERR_MISMATCHED_PAYLOAD,
+            Error::SignatureFailure => ERR_SIGNATURE_FAILURE,
+            Error::Serialisation(_) => ERR_SERIALISATION,
+            Error::String(_) => ERR_STRING,
+            Error::UnknownPeer => ERR_UNKNOWN_PEER,
+            Error::InvalidEvent => ERR_INVALID_EVENT,
+            Error::UnknownParent => ERR_UNKNOWN_PARENT,
+            Error::DuplicateVote => ERR_DUPLICATE_VOTE,
+            Error::Utf8(_) => ERR_UTF8,
+
```
Comment (maqi) [inline src/error.rs]: why a blank line here?

File: src/error.rs [hunk 2]
```diff
@@ -46,10 +54,64 @@ quick_error! {
...
         }
+        /// Unexpected error. Contains custom error message.
+        Unexpected(error: String) {
+            description("An unexpected error has occurred.")
+            display("Unexpected error: {}", error)
+        }
+    }
+}
+
+#[allow(missing_docs)]
```
Comment (maqi) [inline src/error.rs]: exception shall be avoided?

File: src/ffi/block.rs [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use ffi::*;
+    use mock;
+    use std::{mem, ptr, slice};
+
+    #[test]
+    fn ffi_block_new() {
```
Comment (maqi) [inline src/ffi/block.rs]: better have a short explaination of what is this test about: the test scenario and the expected result.
same comment to other tests.

File: src/ffi/error.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            let _ = CString::from_raw(err.description as *mut _);
+        }
+        *last.borrow_mut() = None;
+    });
+}
+
+/// Sets the error. For internal use.
+pub(crate) fn err_set(error_code: i32, description: CString) {
+    // Clear the last error.
+    if let Some(err) = err_take() {
```
Comment (maqi) [inline src/ffi/error.rs]: why not just call `err_clear` ?

File: src/ffi/id.rs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+use ffi::utils;
+use id::Proof as NativeProof;
+use mock::PeerId;
+use std::{slice, str};
+
+/// Secret peer signing key.
+///
+/// Should be deallocated with `secret_id_free`.
+pub struct SecretId(pub(crate) PeerId);
+/// Public peer signing key.
```
Comment (maqi) [inline src/ffi/id.rs]: better have a blank line here?

File: src/ffi/id.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+///
+/// Should be deallocated with `public_id_free`.
+pub struct PublicId(pub(crate) PeerId);
+
+/// Creates a public ID from raw bytes pointed by `id` with a size `id_len`.
```
Comment (maqi) [inline src/ffi/id.rs]: there is only `bytes` and `bytes_len` below?

File: src/ffi/id.rs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    utils::catch_unwind_err_set(|| -> Result<_, Error> {
+        let public_id = slice::from_raw_parts(bytes, bytes_len);
+        let peer_id = PeerId::new(str::from_utf8(public_id)?);
+
+        *o_public_id = Box::into_raw(Box::new(PublicId(peer_id)));
+        Ok(())
+    })
+}
+
+/// Returns the bytes contained in `public_id`.
```
Comment (maqi) [inline src/ffi/id.rs]: the `bytes` and `bytes_len` are not returned, but passed back by setting `o_bytes` and `o_bytes_len` ?

File: src/ffi/id.rs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+/// Deallocates a public id.
+#[no_mangle]
+pub unsafe extern "C" fn public_id_free(public_id: *const PublicId) -> i32 {
+    utils::catch_unwind_err_set(|| -> Result<_, Error> {
+        let _ = Box::from_raw(public_id as *mut PublicId);
+        Ok(())
+    })
+}
+
+/// Creates a secret ID from raw bytes pointed by `id` with a size `id_len`.
```
Comment (maqi) [inline src/ffi/id.rs]: only `bytes` and `bytes_len` below

File: src/ffi/id.rs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+    utils::catch_unwind_err_set(|| -> Result<(), Error> {
+        let public_id = slice::from_raw_parts(bytes, bytes_len);
+        let peer_id = PeerId::new(str::from_utf8(public_id)?);
+
+        *o_secret_id = Box::into_raw(Box::new(SecretId(peer_id)));
+        Ok(())
+    })
+}
+
+/// Returns the bytes contained in `secret_id`.
```
Comment (maqi) [inline src/ffi/id.rs]: the `bytes` and `bytes_len` are not returned, but passed back by setting `o_bytes` and `o_bytes_len` ?

File: src/ffi/id.rs [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+/// `o_public_id` must be freed using `public_id_free`.
+#[no_mangle]
+pub unsafe extern "C" fn public_id_from_bytes(
+    bytes: *const u8,
+    bytes_len: usize,
+    o_public_id: *mut *const PublicId,
+) -> i32 {
+    utils::catch_unwind_err_set(|| -> Result<_, Error> {
+        let public_id = slice::from_raw_parts(bytes, bytes_len);
+        let peer_id = PeerId::new(str::from_utf8(public_id)?);
```
Comment (hunterlester) [inline src/ffi/id.rs]: Is `public_id_from_bytes` meant solely for a mock environment?
I'm here learning how to create a public ID for a peer and noticed that `PeerId` is is declared in `mock.rs`.

Comment (nbaksalyar) [reply src/ffi/id.rs]: @hunterlester,
> Is public_id_from_bytes meant solely for a mock environment?
Not the function per se, but it's current implementation in this PR. Currently, Safe-Crypto is not integrated with PARSEC FFI but the intention is to replace `mock::PeerId` with actual crypto primitives (you can assume that the API won't be changed).

**Comment by pierrechevalier83** [comment]
This work was de-prioritized, so it doesn't look like it will be picked up anytime soon. Closing until the timing is better.

**Pull request closed without merging**